### DISCUSSION TITLE
🦕 Deno compatibility

### DISCUSSION THREAD

**Discussion by gr2m** [octokit/octokit.js#1708] [General] [answered]
Deno v1.0.0 is planned to be released tomorrow:
https://deno.land/

I have experimentend with early versions of Deno and run into lots of trouble. I'll give it another try, with the hope to make as many Octokit libraries compatible with Deno as possible, as fast as possible. I'll keep you all posted in this thread

**Comment by gr2m** [comment]
Currently fighting with problems when importing `@octokit/types`:
https://mobile.twitter.com/gr2m/status/1260304096134717441

**Reply by gr2m** [reply]
That has been resolved 👍

**Comment by gr2m** [comment]
Now experiencing an import problem with `@octokit/graphql`

```
GITHUB_TOKEN=[REDACTED] deno run --allow-env https://gr2m.github.io/deno-playground/octokit-graphql.ts
error: Uncaught TypeError: Cannot read property 'request' of null
const graphql$1 = withDefaults(request.request, {
                                       ^
    at https://cdn.pika.dev/-/@octokit/graphql@v4.3.1-TAZh5RtB2XsWS7QPFqhM/dist=es2019/graphql.js:80:40
    at createCommonjsModule (https://cdn.pika.dev/-/@octokit/graphql@v4.3.1-TAZh5RtB2XsWS7QPFqhM/dist=es2019/graphql.js:9:35)
    at https://cdn.pika.dev/-/@octokit/graphql@v4.3.1-TAZh5RtB2XsWS7QPFqhM/dist=es2019/graphql.js:12:16
```

`request` is imported from https://cdn.pika.dev/-/@octokit/request@v5.4.2-A95EbRgWS0sOyFD7CQLM/dist=es2017/request.js via 

```js
import { request } from '/-/@octokit/request@v5.4.2-A95EbRgWS0sOyFD7CQLM/dist=es2017/request.js';
```

I don't get why `request.request` is null when run with Deno. The import of `@octokit/endpoint` into `@octokit/request` is working the same way but there is no complaint there 🤔

**Reply by kevinkassimo** [reply]
From https://cdn.pika.dev/-/@octokit/request@v5.4.2-A95EbRgWS0sOyFD7CQLM/dist=es2017/request.js :

```js
export { request };
export default null;  // <-- Seems graphql.js is trying to import default instead of import { request }
```

**Reply by gr2m** [reply]
Yes, but it seems that all the builds do the same thing, here is `@octokit/endpoint`: https://cdn.pika.dev/-/@octokit/endpoint@v6.0.1-KJ3wBotRsyOOYTXCSPhA/dist=es2017/endpoint.js

```js
const endpoint = withDefaults(null, DEFAULTS);
export { endpoint };
export default null;
```

And yet, running a script importing `@octokit/request` works without a problem:
deno run --allow-net https://gr2m.github.io/deno-playground/octokit-request.ts

I couldn't find a difference between the two builds ...

**Reply by kevinkassimo** [reply]
`@octokit/request` seems to do `export * from ...` for forwarding named exports, so `import { request } from ...` definitely work. `import request from ...` for default exports should still be `null` though.

**Comment by gr2m** [comment, marked as answer]
Looks like it just works when using cdn.skypack.dev instead of cdn.pika.dev 🎉 

```ts
// index.ts
import { Octokit } from "https://cdn.skypack.dev/@octokit/rest";

console.log(`Octokit.VERSION: ${Octokit.VERSION}`);

const octokit = new Octokit({
  auth: "[personall access token here]",
});

octokit.users.getAuthenticated().then(console.log, console.error);
```

Then run 

```
deno run --allow-net index.ts
```

**Reply by wangkexiong** [reply]
error: TS2339 [ERROR]: Property 'user' does not exist on type 'OctokitWithDefaults'.

**Reply by gr2m** [reply]
it's `users`, not `user`

**Comment by MKruschke** [comment]
@gr2m your example doesn't seem to work anymore. It seems that the types were missing or not correct when I try it right now.

This is what get as output from the cmd-line 
<img width="609" alt="test_ts_—_dependency-discovery-v2__Workspace_" src="https://user-images.githubusercontent.com/1651782/102596419-6b975600-4119-11eb-8337-7a0400ace80b.png">


And so it looks in code
![image](https://user-images.githubusercontent.com/1651782/102596214-2541f700-4119-11eb-91d7-b76ccfba8801.png)

Would love to use that awesome lib also on deno side so maybe you can help out.

**Reply by gr2m** [reply]
hmm it works for me with version 1.2.2, but now fails with 1.6.0 😢 I don't think the problem is with Octokit or Skypack, something changed between these versions. Could you try out at which version it started to fail?

**Reply by MKruschke** [reply]
Hey, just tried different versions of deno. It stoped working with deno v1.5.0.

**Reply by gr2m** [reply]
It also looks like using `@octokit/core` works, but not `@octokit/rest`.

```js
// index.ts
import { Octokit } from "https://cdn.skypack.dev/@octokit/core";

console.log(`Octokit.VERSION: ${Octokit.VERSION}`);

const octokit = new Octokit({
  auth: "[token here]",
});

octokit.request('GET /user').then(console.log, console.error);
```

`@octokit/rest` is actually only a few lines of code, combining `@octokit/rest` with some plugins:
https://github.com/octokit/rest.js/blob/master/src/index.ts

It looks like loading and combining all the libs work, it's just that the REST API endpoint methods from `@octokit/plugin-rest-endpoint-methods` plugin do not get set correctly. Or to be more precise, deno is not able to parse the types, which are somewhat more elaborate

If you replace 

```diff
- octokit.users.getAuthenticated().then(console.log, console.error);
+ octokit.request('GET /user').then(console.log, console.error);
```

Then deno does not throw any errors.

The problem is the same for all plugins that add more methods to the `octokit` instance. E.g. 

```js
octokit.paginate('GET /notifications').then(console.log, console.error);
```

Also throws the same error.

Here is the TypeScript code which makes the plugin mechanism work, it's rather hacky, and it took a long time to make it work, I'd prefer not to touch it: https://github.com/octokit/core.js/blob/342fbdfec007f7737bd8985b608e90da73f6a544/src/index.ts#L51-L70

Maybe you can experiment with it and also compare the release notes of Deno 1.5 to see what the cause might be? I'd very much appreciate your help with this problem

**Reply by gr2m** [reply]
It might help to debug this problem with https://github.com/gr2m/javascript-plugin-architecture-with-typescript-definitions. This is a separate repository which only has the code for the plugin architecture that Octokit is build upon. There is one difference that I didn't get to implement yet, there is an open PR here: https://github.com/gr2m/javascript-plugin-architecture-with-typescript-definitions/pull/15, I need to change `Base.plugin([plugin1, plugin2])` to `Base.plugin(plugin1, plugin2)` because of reasons linked here: https://github.com/gr2m/javascript-plugin-architecture-with-typescript-definitions/pull/15#issuecomment-603426926

**Reply by MKruschke** [reply]
@gr2m skypack also offers to load [automatically type declarations](https://docs.skypack.dev/skypack-cdn/code/deno#automatic-typescript-declarations) by appending `?dts` to the url, but when I try to run it then it fails by fetching @octokit/openapi-types@v2.0.0 

![image](https://user-images.githubusercontent.com/1651782/102721707-d359c680-42fc-11eb-90bd-883a4574b2e7.png)

**Reply by MKruschke** [reply]
Just tried your example with the core module. Execution is working, but vscode doesn't provide intellisense neither on the class Octokit nor on the instance.

**Reply by MKruschke** [reply]
Hey Gregor, I tried **jspm.dev** instead of **skypack** and this doesn't throw an error when I run the rest wrapper. The type of Octokit is `any` so it seems the type definitions are not fetch correctly. 

![image](https://user-images.githubusercontent.com/1651782/102723631-8b8d6c00-4309-11eb-8c73-d062e9ad213b.png)

**Reply by gr2m** [reply]
> skypack also offers to load [automatically type declarations](https://docs.skypack.dev/skypack-cdn/code/deno#automatic-typescript-declarations) by appending `?dts` to the url, but when I try to run it then it fails by fetching @octokit/openapi-types@v2.0.0

`@octokit/openapi-types@v2.0.0` does not export any code, only types. Might that be the problem? See the published package at https://unpkg.com/browse/@octokit/openapi-types@2.0.0/

Do the types work for other packages, so we could compare? All the types do work with intellisense in VS Code for both, TypeScript and JavaScript code. And they work when building with `tsc`

**Reply by gr2m** [reply]
ah wait looks like that package does not have the `"types"` key in its package.json, let me fix that

**Reply by gr2m** [reply]
fixed in https://github.com/octokit/openapi-types.ts/releases/tag/v2.0.1, looks like skypack has it already: https://cdn.skypack.dev/@octokit/openapi-types?dts

```
export * from '/new/@octokit/openapi-types@v2.0.1/dist=es2020';
export {default} from '/new/@octokit/openapi-types@v2.0.1/dist=es2020';
```

**Reply by gr2m** [reply]
Code still fails for me, it still tries to load `@octokit/openapi-types@v2.0.0`. I don't know if it's cache by deno or skypack, do you have any idea?

**Reply by gr2m** [reply]
When I try to run a file with only this line

```ts
import { Endpoints } from "https://cdn.skypack.dev/@octokit/types?dts";
```

Then `deno` loads the latest versions of `@octokit/types` and `@octokit/openapi-types`, but still fails

```
Download https://cdn.skypack.dev/@octokit/types?dts
Download https://cdn.skypack.dev/-/@octokit/types@v6.1.2-PAe7WSg8t9svLCFM4LW1/dist=es2020/@octokit/types.js
Download https://cdn.skypack.dev/-/@octokit/types@v6.1.2-PAe7WSg8t9svLCFM4LW1/dist=es2020,mode=types/dist-types/index.d.ts
Download https://cdn.skypack.dev/-/@octokit/types@v6.1.2-PAe7WSg8t9svLCFM4LW1/dist=es2020,mode=types/dist-types/RequestMethod.d.ts
Download https://cdn.skypack.dev/-/@octokit/types@v6.1.2-PAe7WSg8t9svLCFM4LW1/dist=es2020,mode=types/dist-types/ResponseHeaders.d.ts
Download https://cdn.skypack.dev/-/@octokit/types@v6.1.2-PAe7WSg8t9svLCFM4LW1/dist=es2020,mode=types/dist-types/EndpointOptions.d.ts
Download https://cdn.skypack.dev/-/@octokit/types@v6.1.2-PAe7WSg8t9svLCFM4LW1/dist=es2020,mode=types/dist-types/RequestHeaders.d.ts
Download https://cdn.skypack.dev/-/@octokit/types@v6.1.2-PAe7WSg8t9svLCFM4LW1/dist=es2020,mode=types/dist-types/generated/Endpoints.d.ts
Download https://cdn.skypack.dev/-/@octokit/types@v6.1.2-PAe7WSg8t9svLCFM4LW1/dist=es2020,mode=types/dist-types/OctokitResponse.d.ts
Download https://cdn.skypack.dev/-/@octokit/types@v6.1.2-PAe7WSg8t9svLCFM4LW1/dist=es2020,mode=types/dist-types/RequestInterface.d.ts
Download https://cdn.skypack.dev/-/@octokit/types@v6.1.2-PAe7WSg8t9svLCFM4LW1/dist=es2020,mode=types/dist-types/Signal.d.ts
Download https://cdn.skypack.dev/-/@octokit/types@v6.1.2-PAe7WSg8t9svLCFM4LW1/dist=es2020,mode=types/dist-types/Url.d.ts
Download https://cdn.skypack.dev/-/@octokit/types@v6.1.2-PAe7WSg8t9svLCFM4LW1/dist=es2020,mode=types/dist-types/RequestRequestOptions.d.ts
Download https://cdn.skypack.dev/-/@octokit/types@v6.1.2-PAe7WSg8t9svLCFM4LW1/dist=es2020,mode=types/dist-types/Fetch.d.ts
Download https://cdn.skypack.dev/-/@octokit/types@v6.1.2-PAe7WSg8t9svLCFM4LW1/dist=es2020,mode=types/dist-types/EndpointDefaults.d.ts
Download https://cdn.skypack.dev/-/@octokit/types@v6.1.2-PAe7WSg8t9svLCFM4LW1/dist=es2020,mode=types/dist-types/Route.d.ts
Download https://cdn.skypack.dev/-/@octokit/types@v6.1.2-PAe7WSg8t9svLCFM4LW1/dist=es2020,mode=types/dist-types/AuthInterface.d.ts
Download https://cdn.skypack.dev/-/@octokit/types@v6.1.2-PAe7WSg8t9svLCFM4LW1/dist=es2020,mode=types/dist-types/StrategyInterface.d.ts
Download https://cdn.skypack.dev/-/@octokit/types@v6.1.2-PAe7WSg8t9svLCFM4LW1/dist=es2020,mode=types/dist-types/VERSION.d.ts
Download https://cdn.skypack.dev/-/@octokit/types@v6.1.2-PAe7WSg8t9svLCFM4LW1/dist=es2020,mode=types/dist-types/EndpointInterface.d.ts
Download https://cdn.skypack.dev/-/@octokit/types@v6.1.2-PAe7WSg8t9svLCFM4LW1/dist=es2020,mode=types/dist-types/RequestOptions.d.ts
Download https://cdn.skypack.dev/-/@octokit/types@v6.1.2-PAe7WSg8t9svLCFM4LW1/dist=es2020,mode=types/dist-types/RequestError.d.ts
Download https://cdn.skypack.dev/-/@octokit/types@v6.1.2-PAe7WSg8t9svLCFM4LW1/dist=es2020,mode=types/dist-types/GetResponseTypeFromEndpointMethod.d.ts
Download https://cdn.skypack.dev/-/@octokit/types@v6.1.2-PAe7WSg8t9svLCFM4LW1/dist=es2020,mode=types/dist-types/RequestParameters.d.ts
Download https://cdn.skypack.dev/new/@octokit/openapi-types@v2.0.1/dist=es2020,mode=types?from=@octokit/types
Download https://cdn.skypack.dev/error/unknown:http?from=@octokit/types
Download https://cdn.skypack.dev/-/@octokit/openapi-types@v2.0.1-GTTxr3sBUIWafls76yKQ/dist=es2020/@octokit/openapi-types.js
error: Import 'https://cdn.skypack.dev/-/@octokit/openapi-types@v2.0.1-GTTxr3sBUIWafls76yKQ/dist=es2020/@octokit/openapi-types.js' failed: 404 Not Found
    at https://cdn.skypack.dev/-/@octokit/types@v6.1.2-PAe7WSg8t9svLCFM4LW1/dist=es2020,mode=types/dist-types/generated/Endpoints.d.ts:1:0
```

**Reply by gr2m** [reply]
@G-Rath [because you offered to help with TypeScript stuff](https://github.com/octokit/webhooks/pull/254#issuecomment-757088740) :) Have you played around with Deno before? There is some discrepancy here. `@octokit/rest`'s Type definitions work with the Intellinse in VS Code, but Deno seems to have a problem. It might also be related to Skypeack, I'm not sure. I've found that importing `@octokit/rest` into TypeScripts playground also doesn't load the Type Definitions ([playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgeQMYwga2PAvnAMyghDgCIABCdLHAeigFMBnGMgKHdQgDtW5qGbPAC8cHowDuKGsIAUASk6DaMAHRNIzNQHNGMRUA)). It would be amazing if we could both to work, but even after working over a year with TypeScript, I'm still having a very hard time debugging it, and this seems to be a rather complex problem. Maybe you have an idea of what might go wrong, or how to approach the problem?

**Reply by G-Rath** [reply]
I've not used Deno before, but happy to take a look :)

I think both Deno and playground are failing for the same reason: they're looking for types in `dist-types/generated`:

> error: Import 'https://cdn.skypack.dev/-/@octokit/openapi-types@v2.0.1-GTTxr3sBUIWafls76yKQ/dist=es2020/@octokit/openapi-types.js' failed: 404 Not Found
    at https://cdn.skypack.dev/-/@octokit/types@v6.1.2-PAe7WSg8t9svLCFM4LW1/dist=es2020,mode=types/dist-types/generated/Endpoints.d.ts:1:0

![image](https://user-images.githubusercontent.com/3151613/104108533-5681a180-532a-11eb-9032-194e5986a49f.png)

I'll try and have a look sometime today and let you know if I can figure out what's going on.

**Reply by G-Rath** [reply]
🤦 I'm pretty sure is a bug in a parser somewhere - it's seeing `"DELETE /repos/{owner}/{repo}/import": Operation<"/repos/{owner}/{repo}/import", "delete">;` and thinking `import: Operation<"/repos/{owner}/{repo}/import", "delete` .

**Reply by G-Rath** [reply]
Ironically this seems to have been [fixed 3 days ago](https://github.com/microsoft/TypeScript-Website/commit/c64d6d37b6ad05c074969beb2dbb9d8831e5e39b) due to a change in the regexp, which means it no longer matches this string.

I'm guessing Deno uses the same old regexp? I'll see if I can figure out the fix for that.

This is the PR: https://github.com/microsoft/TypeScript-Website/pull/1470

**Reply by gr2m** [reply]
It looks like the fix is automatically deployed to https://www.staging-typescript.org, I don't see any `404` requests ([playground](https://www.staging-typescript.org/play?ssl=8&ssc=1&pln=8&pc=20#code/JYWwDg9gTgLgBAbzgeQMYwga2PAvnAMyghDgCIABCdLHAeigFMBnGMgKFElkRRu3gBDZnADC0RnHxES5Kv3qoJHdkoB2rONQwDxTOAF44axgHcxEgBQBKVRA3xtteEZPm0OnDfbsnuiQB0TACOAK4sMJZkAOIAogAqcHRktn44QYyQzAEA5oyR1kA)). But unfortunately the types are still not loaded for `@octokit/rest`

**Reply by G-Rath** [reply]
Yeah they're deployed every Monday.

Now I think the issue is the size of the types means it's not evaluating them. Maybe @orta has some ideas?

**Reply by orta** [reply]
Good spot, yeah, this definitely used to work because I used this dep for testing transitive imports (the npm module danger exposes an auth'd GH rest client) - the playground ATA is a pretty solid series of global hacks, so it's quite possible that something gets lost in there (lodash doesn't work either)

![Screen Shot 2021-01-13 at 8 50 47 AM](https://user-images.githubusercontent.com/49038/104428613-9770f480-557c-11eb-8d8a-7c744842240f.png)

It's been on my TODO to try move to pre-downloading all of it and letting the real ATA in TypeScript do the work. Consider this now higher on my TODO.

**Reply by gr2m** [reply]
Thank you Orta, appreciate your help!

**Reply by gr2m** [reply]
I fixed a few problems that likely affected both `deno` and the TypeScript playground.

I got the following working now with `deno`

```ts
import { Octokit } from "https://cdn.skypack.dev/@octokit/core?dts";
import { restEndpointMethods } from "https://cdn.skypack.dev/@octokit/plugin-rest-endpoint-methods?dts";

const MyOctokit = Octokit.plugin(restEndpointMethods);

const octokit = new MyOctokit({
  auth: Deno.env.get("GITHUB_TOKEN"),
});

const { data } = await octokit.repos.get({
  owner: "gr2m",
  repo: "sandbox",
});

console.log(data);
```

I get full IntelliSense support in VS Code with the [deno extension](https://marketplace.visualstudio.com/items?itemName=denoland.vscode-deno) :tada:

However, when trying to use `@octokit/rest` I get an error

```ts
import { Octokit } from "https://cdn.skypack.dev/@octokit/rest?dts";

const octokit = new Octokit({
  auth: Deno.env.get("GITHUB_TOKEN"),
});

const { data } = await octokit.repos.get({
  owner: "gr2m",
  repo: "sandbox",
});

console.log(data);
```

> ```
> error: TS2694 [ERROR]: Namespace '__' has no exported member 'RestEndpointMethods'.
>    at https://cdn.skypack.dev/-/@octokit/rest@v18.1.1-RlR1JwzLQt2h84FvZXCC/dist=es2020,mode=types/dist-types/index.d.ts:12:147
>```

I don't understand what the `'__'` namespace is, I don't see it in the source code.

@orta It's not quite working in the TypeScript playground yet ([playground link](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgeQMYwga2PAvnAMyghDgCIABCdLHAelWgFMyBuAKFElkTjAEMA5sAB2-GEwBKTAM55CxUpWoZsMOmAA2AV2EiAtAL3im+qLJht27RiLlwVteAF44IpgHcUNNQAoAlNZ0dHAAkiISmprAAMpMdkyE0A4+OAB05gCO2hZwHtCYMgA07MFwAEba8CIQ8DAAFokwAJ5gTPyN-AAmSVBwDYnEVYkCUPwgTBJQNhB28Ehd4vxw+K78Hvw4KarpWTlyvmQA4gCiACpwdGSBtjIQmkxpmhCCvosw-IEzc947MADqOHqAAUdHo4K40H80lpdKJfEZRCZpHJ-BxbvBHGpAQ1gUIkRIIW5PL8nDiQWD4WjrBi4MAZDJ9kT1ptMakAUC8cYJDD8WIJIdThc6OZIDI6AgIB53FBcBLRRA5fTGbIyEVEOw4A5pUwoAAuciCKAAJhAas1cAVBrIMn4Ii65QgAA9zVq5CZrRA2iJzbhqbd7o9nq9lfs0UA))

**Reply by gr2m** [reply]
> It also looks like using `@octokit/core` works, but not `@octokit/rest`.

you can do the following to get the same as importing `@octokit/rest`

```ts
import { Octokit as Core } from "https://cdn.skypack.dev/@octokit/core?dts";
import { requestLog } from "https://cdn.skypack.dev/@octokit/plugin-request-log?dts";
import { paginateRest } from "https://cdn.skypack.dev/@octokit/plugin-paginate-rest?dts";
import { restEndpointMethods } from "https://cdn.skypack.dev/@octokit/plugin-rest-endpoint-methods?dts";

const Octokit = Core.plugin(
  requestLog,
  restEndpointMethods,
  paginateRest
)
```

And that works without a problem 🤷🏼

**Reply by orta** [reply]
The playground issues are fixed upstream in the playground https://github.com/microsoft/TypeScript-Website/commit/ac2c165726f81cbcc9387f079d56b54a4cd3c106 and should get deployed this week ( that example octokit playground [on twitter](https://twitter.com/orta/status/1361331963290918916) was because this idea had been nagging me for a few months )

![image](https://user-images.githubusercontent.com/49038/108032530-89eed480-702a-11eb-991d-8cc80fbdbfef.png)

**Reply by gr2m** [reply]
> The playground issues are fixed upstream in the playground [microsoft/TypeScript-Website@ac2c165](https://github.com/microsoft/TypeScript-Website/commit/ac2c165726f81cbcc9387f079d56b54a4cd3c106) and should get deployed this week

Yay! It seems to work now! 🎉 [playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgeQMYwga2PAvnAMyghDgCIABCdLHAeigFMBnGMgbgChUIA7VuNQzZ4AXji9GAdxQ0RACgCUXbnwFIAJgEMYWuPnFapWnILk4AdE0jMLAc0Yx5CTnEFTJUAFzk7UAEwgZAA0rnDWED5kzFq8GgBGEAAeIZy4ypyZWdk5PPwQADaMFgUQdvLauopAA)

<img width="1065" alt="image" src="https://user-images.githubusercontent.com/39992/108468004-efb1ab00-723a-11eb-83d7-6a3228f04d6e.png">

Thanks @orta

**Comment by oscard0m** [comment]
More news from Deno 🦕 : https://deno.com/blog/changes

Let's see how it works with `npm:@octokit/core` in the upcoming months 🍿
